react/features/base/participants/components/AvatarImage.native.js

```diff
@@ -0,0 +1,200 @@
+     * If the URI is undefined, wait {@code UNDEFINED_AVATAR_TIMEOUT} ms and
+     * start showing a default locally generated avatar afterwards.
+     *
+     * Once a URI is passed, it will be rendered instead, except if loading it
+     * fails, in which case we fallback to a locally generated avatar again.
+     *
+     * @inheritdoc
+     * @param {Object} nextProps - The read-only React Component props that this
+     * instance will receive.
+     * @returns {void}
+     */
+    componentWillReceiveProps(nextProps) {
+        const prevURI = this.props.source && this.props.source.uri;
+        const nextURI = nextProps.source && nextProps.source.uri;
+
+        if (typeof prevURI === 'undefined') {
+            clearTimeout(this._timeout);
+            if (typeof nextURI === 'undefined') {
+                this._timeout = setTimeout(() => {
+                    this.setState({ showDefault: true });
+                }, UNDEFINED_AVATAR_TIMEOUT);
+            } else {
+                this.setState({ showDefault: nextProps.forceDefault });
+            }
+        }
+    }
+
+    /**
+     * Clear the timer just in case. See {@code componentWillReceiveProps} for
+     * details.
+     *
+     * @inheritdoc
+     */
+    componentWillUnmount() {
+        clearTimeout(this._timeout);
+    }
+
+    /**
+     * Computes a hash over the URI and returns a HSL background color. We use
+     * 75% as lightness, for nice pastel style colors.
+     *
+     * @returns {string} - The HSL CSS property.
+     * @private
+     */
+    _getBackgroundColor() {
+        const uri = this.props.source.uri;
+        let hash = 0;
+
+        // If we have no URI yet we have no data to hash from, so use a random
+        // value.
+        if (typeof uri === 'undefined') {
+            hash = Math.floor(Math.random() * 360);
+        } else {
+            /* eslint-disable no-bitwise */
+
+            for (let i = 0; i < uri.length; i++) {
+                hash = uri.charCodeAt(i) + ((hash << 5) - hash);
+                hash |= 0;  // Convert to 32bit integer
+            }
+
+            /* eslint-enable no-bitwise */
+        }
+
+        return `hsl(${hash % 360}, 100%, 75%)`;
+    }
+
+    /**
+     * Error handler for image loading. When an image fails to load we'll mark
+     * it as failed and load the default URI instead.
+     *
+     * @private
+     * @returns {void}
+     */
+    _onError() {
+        this.setState({ failed: true });
+    }
+
+    /**
+     * Implements React's {@link Component#render()}.
+     *
+     * @inheritdoc
+     */
+    render() {
+        // eslint-disable-next-line no-unused-vars
+        const { forceDefault, source, style, ...props } = this.props;
+        const { failed, showDefault } = this.state;
+
+        if (failed || showDefault) {
+            const coloredBackground = {
+                ...style,
+                backgroundColor: this._getBackgroundColor(),
```

Comment:
So we plan to generate a random number or compute a hash on any render, regardless of whether the render was absolutely necessary or whether the state from which `_getBackgroundColor` actually derives its value did really change? I'm not comfortable with such an approach. We know computing a hash or generating a random number is  non-trivial, we know it derives data from certain pieces of state so I thin it should be done as soon as and only when the state changes in a way which requires the `backgroundColor` to change. Then `render` uses whatever has been calculated already.